saleor/cart/models.py

```diff
@@ -0,0 +1,232 @@
+
+    def save(self):
+        self.update(status=Cart.SAVED)
+
+
+class Cart(models.Model, ItemSet):
+
+    COOKIE_NAME = 'cart'
+
+    OPEN, SAVED, WAITING_FOR_PAYMENT, ORDERED, CHECKOUT, CANCELED = (
+        'open', 'saved', 'payment', 'ordered', 'checkout', 'canceled')
+
+    STATUS_CHOICES = (
+        (OPEN, pgettext_lazy('Cart', 'Open - currently active')),
+        (WAITING_FOR_PAYMENT, pgettext_lazy('Cart', 'Waiting for payment')),
+        (SAVED, pgettext_lazy(
+            'Cart', 'Saved - for items to be purchased later')),
+        (ORDERED, pgettext_lazy(
+            'Cart', 'Submitted - has been ordered at the checkout')),
+        (CHECKOUT, pgettext_lazy(
+            'Cart', 'Checkout - basket is processed in checkout')),
+        (CANCELED, pgettext_lazy(
+            'Cart', 'Canceled - basket was canceled by user'))
+    )
+
+    status = models.CharField(
+        pgettext_lazy('Cart', 'order status'),
+        max_length=32, choices=STATUS_CHOICES, default=OPEN)
+    created = models.DateTimeField(
+        pgettext_lazy('Cart', 'created'), auto_now_add=True)
+    last_status_change = models.DateTimeField(
+        pgettext_lazy('Cart', 'last status change'), auto_now_add=True)
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, blank=True, null=True, related_name='carts',
+        verbose_name=pgettext_lazy('Cart', 'user'))
+    email = models.EmailField(blank=True, null=True)
+    token = models.UUIDField(pgettext_lazy('Cart', 'token'),
+                             primary_key=True, default=uuid4, editable=False)
+    voucher = models.ForeignKey(
+        Voucher, null=True, related_name='+', on_delete=models.SET_NULL)
+    checkout_data = JSONField(null=True, editable=False)
+
+    total = PriceField(
+        currency=settings.DEFAULT_CURRENCY, max_digits=12, decimal_places=2,
+        default=0)
+    quantity = models.PositiveIntegerField(default=0)
+
+    objects = CartQueryset.as_manager()
+
+    class Meta:
+        ordering = ('-last_status_change',)
+
+    def __init__(self, *args, **kwargs):
+        self.discounts = kwargs.pop('discounts', None)
+        super(Cart, self).__init__(*args, **kwargs)
+
+    def update_quantity(self):
+        total_lines = self.count()['total_quantity']
+        if not total_lines:
+            total_lines = 0
+        self.quantity = total_lines
+        self.save(update_fields=['quantity'])
+
+    def change_status(self, status):
+        if status not in dict(self.STATUS_CHOICES):
+            raise ValueError('Not expected status')
+        if status != self.status:
+            self.status = status
+            self.last_status_change = now()
+            self.save()
+
+    def is_shipping_required(self):
+        return any(line.is_shipping_required() for line in self.lines.all())
+
+    #Satchless
+    def __repr__(self):
+        return 'Cart(%r)' % (list(self),)
+
+    def __iter__(self):
+        return iter(self.lines.select_related('product').all())
```

Comment:
By adding `select_related` here you can force a new DB query to be executed even if lines have already been prefetched.